miner/scroll_worker.go

```diff
@@ -0,0 +1,945 @@
+// Note the worker does not support being closed multiple times.
+func (w *worker) close() {
+	atomic.StoreInt32(&w.running, 0)
+	close(w.exitCh)
+	w.wg.Wait()
+}
+
+// newWorkLoop is a standalone goroutine to submit new mining work upon received events.
+func (w *worker) newWorkLoop(recommit time.Duration) {
+	defer w.wg.Done()
+	var (
+		timestamp int64 // timestamp for each round of mining.
+	)
+
+	// commit aborts in-flight transaction execution with given signal and resubmits a new one.
+	commit := func(noempty bool) {
+		select {
+		case w.newWorkCh <- &newWorkReq{noempty: noempty, timestamp: timestamp}:
+		case <-w.exitCh:
+			return
+		}
+		atomic.StoreInt32(&w.newTxs, 0)
+		atomic.StoreInt32(&w.newL1Msgs, 0)
+	}
+	// clearPending cleans the stale pending tasks.
+	clearPending := func(number uint64) {
+		w.pendingMu.Lock()
+		for h, t := range w.pendingTasks {
+			if t.block.NumberU64()+staleThreshold <= number {
+				delete(w.pendingTasks, h)
+			}
+		}
+		w.pendingMu.Unlock()
+	}
+
+	for {
+		select {
+		case <-w.startCh:
+			clearPending(w.chain.CurrentBlock().NumberU64())
+			timestamp = time.Now().Unix()
+			commit(false)
+		case head := <-w.chainHeadCh:
+			clearPending(head.Block.NumberU64())
+			timestamp = time.Now().Unix()
+			commit(true)
+		case <-w.exitCh:
+			return
+		}
+	}
+}
+
+// mainLoop is a standalone goroutine to regenerate the sealing task based on the received event.
+func (w *worker) mainLoop() {
+	defer w.wg.Done()
+	defer w.txsSub.Unsubscribe()
+	defer w.l1MsgsSub.Unsubscribe()
+	defer w.chainHeadSub.Unsubscribe()
+	defer w.chainSideSub.Unsubscribe()
+
+	deadCh := make(chan *pipeline.Result)
+	pipelineResultCh := func() <-chan *pipeline.Result {
+		if w.currentPipeline == nil {
+			return deadCh
+		}
+		return w.currentPipeline.ResultCh
+	}
+
+	for {
+		select {
+		case req := <-w.newWorkCh:
+			w.startNewPipeline(req.timestamp)
+		case result := <-pipelineResultCh():
+			w.handleCCCResult(result)
+		case ev := <-w.txsCh:
+			// Apply transactions to the pending state
+			//
+			// Note all transactions received may not be continuous with transactions
+			// already included in the current mining block. These transactions will
+			// be automatically eliminated.
+			if w.currentPipeline != nil && w.currentPipeline.TxnQueue != nil {
+				txs := make(map[common.Address]types.Transactions)
+				signer := types.MakeSigner(w.chainConfig, w.currentPipeline.Header.Number)
+				for _, tx := range ev.Txs {
+					acc, _ := types.Sender(signer, tx)
+					txs[acc] = append(txs[acc], tx)
+				}
+				txset := types.NewTransactionsByPriceAndNonce(signer, txs, w.currentPipeline.Header.BaseFee)
+				if result := w.currentPipeline.TryPushTxns(txset, w.onTxFailingInPipeline); result != nil {
+					w.handleCCCResult(result)
+				}
+			}
+			atomic.AddInt32(&w.newTxs, int32(len(ev.Txs)))
+
+		case ev := <-w.l1MsgsCh:
+			atomic.AddInt32(&w.newL1Msgs, int32(ev.Count))
```

Comment:
Seems like this is not used anymore?

Reply:
Yeah, but that was the case for the existing worker as well. Will remove regardless.

Reply:
https://github.com/scroll-tech/go-ethereum/pull/735/commits/62e71907daf43aaea87432656dbe70c9f2b70a23